internal/collector/gpu/gpu.go

```diff
@@ -0,0 +1,398 @@
+		"Non-local adapter memory usage in bytes.",
+		[]string{"phys"},
+		nil,
+	)
+
+	c.gpuProcessMemoryDedicatedUsage = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "process_memory_dedicated_bytes"),
+		"Dedicated process memory usage in bytes.",
+		[]string{"pid", "phys"},
+		nil,
+	)
+	c.gpuProcessMemoryLocalUsage = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "process_memory_local_bytes"),
+		"Local process memory usage in bytes.",
+		[]string{"pid", "phys"},
+		nil,
+	)
+	c.gpuProcessMemoryNonLocalUsage = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "process_memory_non_local_bytes"),
+		"Non-local process memory usage in bytes.",
+		[]string{"pid", "phys"},
+		nil,
+	)
+	c.gpuProcessMemorySharedUsage = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "process_memory_shared_bytes"),
+		"Shared process memory usage in bytes.",
+		[]string{"pid", "phys"},
+		nil,
+	)
+	c.gpuProcessMemoryTotalCommitted = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "process_memory_committed_bytes"),
+		"Total committed process memory in bytes.",
+		[]string{"pid", "phys"},
+		nil,
+	)
+
+	c.gpuEnginePerfDataCollector, err = pdh.NewCollector[gpuEnginePerfDataCounterValues](pdh.CounterTypeRaw, "GPU Engine", pdh.InstancesAll)
+	if err != nil {
+		return fmt.Errorf("failed to create GPU Engine perf data collector: %w", err)
+	}
+	c.gpuAdapterMemoryPerfDataCollector, err = pdh.NewCollector[gpuAdapterMemoryPerfDataCounterValues](pdh.CounterTypeRaw, "GPU Adapter Memory", pdh.InstancesAll)
+	if err != nil {
+		return fmt.Errorf("failed to create GPU Adapter Memory perf data collector: %w", err)
+	}
+	c.gpuLocalAdapterMemoryPerfDataCollector, err = pdh.NewCollector[gpuLocalAdapterMemoryPerfDataCounterValues](pdh.CounterTypeRaw, "GPU Local Adapter Memory", pdh.InstancesAll)
+	if err != nil {
+		return fmt.Errorf("failed to create GPU Local Adapter Memory perf data collector: %w", err)
+	}
+	c.gpuNonLocalAdapterMemoryPerfDataCollector, err = pdh.NewCollector[gpuNonLocalAdapterMemoryPerfDataCounterValues](pdh.CounterTypeRaw, "GPU Non Local Adapter Memory", pdh.InstancesAll)
+	if err != nil {
+		return fmt.Errorf("failed to create GPU Non Local Adapter Memory perf data collector: %w", err)
+	}
+	c.gpuProcessMemoryPerfDataCollector, err = pdh.NewCollector[gpuProcessMemoryPerfDataCounterValues](pdh.CounterTypeRaw, "GPU Process Memory", pdh.InstancesAll)
+	if err != nil {
+		return fmt.Errorf("failed to create GPU Process Memory perf data collector: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Collector) Collect(ch chan<- prometheus.Metric) error {
+	type PidPhysEngEngType struct {
+		Pid     string
+		Phys    string
+		Eng     string
+		Engtype string
+	}
+	type PidPhys struct {
+		Pid  string
+		Phys string
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Collect the GPU Engine perf data
```

Comment:
would it be possible to separate the collector into different functions, for a better readability?